[Mobile]Add video block (first iteration)

pinarol opened the pull request:
## Description

To Fix: https://github.com/wordpress-mobile/gutenberg-mobile/issues/855

This PR introduces the new video block for Mobile. First of all I am sorry for this huge PR. I will try to add notes and make it easier to understand as much as possible. Please note that most of the green changes aren't actually new code, they are mostly extracted from image block, I added some inline comments to point this out.   

## How has this been tested?
Tested via: https://github.com/wordpress-mobile/gutenberg-mobile/pull/854

## Types of changes

- Extracted MediaUpload component from Image block (similar to web )
- Extracted MediaUploadProgress component from Image block, it's now been used by image and video blocks
- Updated MediaPlaceholder component to be suitable for different media types ( similar to web )
- Added video block and reused MediaUpload, MediaUploadProgress, MediaPlaceholder components
- Added react-native-video 3rd party library to have the preview of videos. Also this makes it similar to web side which uses html `<video>` tag, so hopefully any future abstractions will be easier to make the video block cross platform. 

I added some additional inline comments to describe what's done.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

pinarol committed:
Pass media type to bridge

pinarol committed:
Add video block

pinarol committed:
Merge branch 'master' into rnmobile/update-media-type

# Conflicts:
#	packages/block-library/src/index.native.js

pinarol committed:
Add media upload ui

pinarol committed:
Merge branch 'master' into rnmobile/update-media-type

# Conflicts:
#	packages/block-library/src/index.native.js

pinarol committed:
Add video block

pinarol committed:
Extract media upload

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

pinarol committed:
Fix lint issues

pinarol committed:
Integrate RN video

pinarol committed:
Integrate RN video

pinarol committed:
Fix lint issues

pinarol committed:
Rename MediaUploadUI

pinarol committed:
Separate Android and iOS video player configs

pinarol committed:
Fix lint issues

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

pinarol committed:
Revert unwanted auto linter fix

pinarol committed:
Update order of video block

pinarol committed:
Fix typo

pinarol commented on packages/block-library/src/image/media-upload-progress.native.js:
```diff
@@ -0,0 +1,151 @@
+/**
```
This is a new component to be used by both video and image blocks(and probably by other media upload blocks in the future). It is actually not a new code, just extracted from image block. 

It listens to media upload events coming from main apps and updates progress Spinner. It also provides a way to render a child component with the current upload state(see props.renderContent) If provided a coverUrl via props then it will also calculate width, height of it and pass to renderContent.

etoledom replied:
Maybe a better place for this component would be `media-upload/`? Since it's meant to be shared between media related blocks.
It also looks like a good candidate for a HOC (not 100% sure). This might also make easier its implementation in blocks.

pinarol replied:
Thanks @etoledom 👍 I added a tech debt here https://github.com/wordpress-mobile/gutenberg-mobile/issues/687#issuecomment-482011555

pinarol commented on packages/block-editor/src/components/media-upload/index.native.js:
```diff
@@ -0,0 +1,98 @@
+/**
```
This is a new component to be used by both video and image blocks also by MediaUpload component. It is actually not a new code, just extracted from image block.

This component is responsible for displaying media options picker to user. It is designed similar to what web has, it passes a function 'open' via its render method and the rendered component decides when to call that function. 

Different thing is web doesn't need to pass `getMediaOptions` but we do, the renderer also needs to embed `getMediaOptions` call to an appropriate place. I tried wrapping this component with a `View` and calling `getMediaOptions` inside, but that messed up the layout of edit button. 🤷‍♀️

pinarol commented on packages/block-editor/src/components/media-placeholder/index.native.js:
```diff
@@ -8,25 +8,67 @@ import { View, Text, TouchableWithoutFeedback } from 'react-native';
  */
 import styles from './styles.scss';
 
 function MediaPlaceholder( props ) {
```
We are making MediaPlaceholder usable by both image and video blocks by passing media type.

pinarol commented on packages/block-editor/src/components/media-placeholder/index.native.js:
```diff
@@ -8,25 +8,67 @@ import { View, Text, TouchableWithoutFeedback } from 'react-native';
-				</Text>
-			</View>
-		</TouchableWithoutFeedback>
+		<MediaUpload
```
We are also using our new MediaUpload component here.

pinarol commented on packages/block-library/src/video/edit.native.js:
```diff
@@ -0,0 +1,226 @@
+							return (
+								<View onLayout={ this.onVideoContanerLayout } style={ { flex: 1 } }>
+									{ showVideo &&
+										<Video
```
By using this video component we are coming close to what web side has.

koke commented on packages/block-library/src/video/video-player.ios.js:
```diff
@@ -0,0 +1,73 @@
+					} }
+					// Using built-in player controls is messing up the layout on iOS.
+					// So we are setting controls=false and adding a play button that
+					// will trigger presentFullscreenPlayer()
```
Sounds good as a workaround for this PR, but let's make sure to track this in another issue for 2.0.

pinarol replied:
I've added a new item to the TODO list here: https://github.com/wordpress-mobile/gutenberg-mobile/issues/687
Unfortunately this is in library level, they are setting the frame (as a workaround for some issue) when controls=true
https://github.com/react-native-community/react-native-video/blob/c30f246f5694452f4f081ff30ac4523dfd89f41d/ios/Video/RCTVideo.m#L593
And by doing that our they are messing up the y position of our page. I couldn't come up with a nice solution myself, I was planning to open an issue to their repo after this is merged. I think they might agree that current way is not very healthy. I can also try to tackle this more and send a PR to the original repo if they ignore the issue in the long run 🤔

pinarol committed:
Enhance placeholder

pinarol committed:
Add retry action for video block

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

marecar3 committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

pinarol committed:
Remove unused state property

marecar3 committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

marecar3 committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

# Conflicts:
#	packages/block-editor/src/components/media-placeholder/index.native.js
#	packages/block-library/src/image/edit.native.js

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

marecar3 committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

# Conflicts:
#	packages/block-editor/src/components/media-placeholder/index.native.js

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

marecar3 committed:
Fixed problem with cancel/retry uploading of video

marecar3 commented on packages/block-library/src/video/edit.native.js:
```diff
@@ -0,0 +1,241 @@
+	componentDidMount() {
+		const { attributes, setAttributes } = this.props;
+		if ( attributes.id && ! isURL( attributes.src ) ) {
+			if ( attributes.url.indexOf( 'file:' ) === 0 ) {
```
Hey @pinarol @SergioEstevao do we need this block of code as it was put for copy/paste images on iOS but not sure for videos?

pinarol replied:
It looks like we don't have support for video copy paste on Aztec either. So yes, we don't need this code, let me remove that.

marecar3 committed:
Set proper bottom sheet icons for video block

marecar3 committed:
Make sure that icon block has latest value of isUploadingInProgress

pinarol committed:
Remove unnecessary code for video copy/paste handling

We are not supporting video copy paste, that code was copy/paste from image block

pinarol committed:
Removing poster support

poster support isn't present on wp.com yet and react-native-video has an ugly bug
about posters so we'll need to solve that first, after that we'll re-add this.

etoledom reviewed (approved):
Thank you for all this great hard work @pinarol !

It looks great and has been more than tested via https://github.com/wordpress-mobile/gutenberg-mobile/pull/854 and https://github.com/wordpress-mobile/WordPress-iOS/pull/11454 (for iOS).

I left a small code commen as a _possible_ future optimization.

pinarol committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

marecar3 committed:
Merge branch 'master' into rnmobile/video-block-first-iteration

marecar3 committed:
Fixed crash when opening the post which previously had cancelled image/video upload

marecar3 commented:
Hey @pinarol, as we have refactored an image and media progress upload in this PR I have manually included this fix: https://github.com/WordPress/gutenberg/pull/15532 into our PR.
